### PR TITLE
Fix clone() method using processed spec instead of raw specFix spec.py

### LINKED ISSUE #2078: Swagger UI endpoint crashes at runtime with JSON Schema keywords like propertyNames or patternProperties
## Description

Connexion fully supports OpenAPI 3.x and injects features like `propertyNames` or `patternProperties` into the generated schema at runtime. These are valid under JSON Schema (especially in OpenAPI 3.1), and Connexion accepts and uses them.

However, Connexion **fails at runtime when serving the Swagger UI**.  
This causes errors like:
```
must NOT have additional properties
must match exactly one schema in oneOf
```

**Concerns:**
- Users expect their runtime schema and documentation schema to align.

## Expected Behavior

Connexion should allow serving Swagger UI for OpenAPI specs containing modern JSON Schema keywords without crashing.

At minimum, it could provide a configurable option to sanitize the spec automatically for the documentation endpoint.

## Steps to Reproduce

1. Create an OpenAPI 3.x spec containing propertyNames or patternProperties.
2. Load it in Connexion with Swagger UI enabled.
3. Access /docs/.
4. Observe runtime error preventing the Swagger UI from loading.

## Environment

- Connexion version: connexion[flask,uvicorn,swagger-ui]==3.2.0
- Python version: 3.12.2

## Our workaround

```py
def patch_openapi_spec(schema):
    """
    Patch OpenAPI spec to replace problematic JSON Schema keywords
    with x-prefixed versions for documentation compatibility
    """
    if isinstance(schema, dict):
        if "patternProperties" in schema:
            schema["x-patternProperties"] = schema.pop("patternProperties")
        if "propertyNames" in schema:
            schema["x-propertyNames"] = schema.pop("propertyNames")
        for v in schema.values():
            patch_openapi_spec(v)
    elif isinstance(schema, list):
        for v in schema:
            patch_openapi_spec(v)

def generate_config_doc_yaml():
    """
    Generate documentation-safe YAML from validation YAML
    """
    validate_config_path = os.path.join(os.path.dirname(__file__), 'openapi', 'config-api.yaml')
    doc_config_path = os.path.join(os.path.dirname(__file__), 'openapi', 'config-api-doc.yaml')
    
    try:
        with open(validate_config_path, 'r', encoding='utf-8') as f:
            spec = yaml.safe_load(f)
        
        patch_openapi_spec(spec)
        
        with open(doc_config_path, 'w', encoding='utf-8') as f:
            yaml.dump(spec, f, sort_keys=False, allow_unicode=True)
            
        return doc_config_path
    except Exception as e:
        logging.error(f"Error generating config doc YAML: {e}")
        return None

app.add_api('config-api.yaml', name='config', arguments={'title': 'Configuration'}, strict_validation=True, swagger_ui=False)

# Custom documentation routes for config API
@app.route("/api/v1.0/configuration/docs/openapi.yaml")
def config_docs_spec():
    """Serve the documentation-safe OpenAPI spec for config API"""
    return send_from_directory(os.path.join(os.path.dirname(__file__), 'openapi'), 'config-api-doc.yaml')

@app.route("/api/v1.0/configuration/docs/")
def config_swagger_ui():
    """Serve custom Swagger UI for config API"""
    return f"""
    <!DOCTYPE html>
    <html>
    <head>
      <title>C3PO Configuration API Documentation</title>
      <link rel="stylesheet" type="text/css" href="https://unpkg.com/swagger-ui-dist/swagger-ui.css" />
    </head>
    <body>
      <div id="swagger-ui"></div>
      <script src="https://unpkg.com/swagger-ui-dist/swagger-ui-bundle.js"></script>
      <script>
      const ui = SwaggerUIBundle({{
        url: '/api/v1.0/configuration/docs/openapi.yaml',
        dom_id: '#swagger-ui',
        deepLinking: true,
        presets: [
          SwaggerUIBundle.presets.apis,
          SwaggerUIBundle.presets.standalone
        ]
      }});
      </script>
    </body>
    </html>
    """

```

### Maybe related

https://github.com/spec-first/connexion/issues/1396
https://github.com/spec-first/connexion/pull/2043
https://editor-next.swagger.io/

### PR DESCRIPTION
**Pull request by giuliohome**
## Problem
When Swagger UI accesses `/docs/` endpoint, it calls `with_base_path()` which triggers `clone()`. The current implementation uses `copy.deepcopy(self._spec)` (processed spec) instead of the original raw specification, causing validation failures that don't occur during app startup.

## Root Cause
The `clone()` method at line 207 creates a new instance from `self._spec` (already processed and resolved) rather than `self._raw_spec` (original). The `deepcopy` operation can corrupt internal references and circular dependencies in the processed spec.

## Fix
```python
# Before
def clone(self):
    return type(self)(copy.deepcopy(self._spec))

# After  
def clone(self):
    return type(self)(copy.deepcopy(self._raw_spec))
```

Fixes #2078

### CODE REVIEW

**Comment by giuliohome** [comment]
# Why Validating Processed Specs is Conceptually Wrong

**Technical Analysis by Claude AI**

## The Core Issue

The `clone()` method validates a **processed specification** (`_spec`) instead of the **original specification** (`_raw_spec`). This is fundamentally flawed because:

## What happens during `resolve_refs()`

When Connexion processes a raw OpenAPI spec, it:

1. **Resolves all `$ref` references** 
2. **Inlines referenced schemas** directly into the spec
3. **Changes the structural context** of schema elements

## Why this breaks validation

### Example: Valid Raw Spec
```yaml
# This passes OpenAPI 3.0 validation
additionalProperties:
  $ref: '#/components/schemas/SomeSchema'
  propertyNames:           # Ignored - not part of $ref validation
    pattern: '^[a-z]+$'
```

### After `resolve_refs()` Processing
```yaml
# This FAILS OpenAPI 3.0 validation  
additionalProperties:
  type: object              # Inlined from resolved $ref
  properties: {...}         # Inlined schema content
  propertyNames:            # Now INVALID - OpenAPI 3.0 doesn't support this
    pattern: '^[a-z]+$'
```

## The Conceptual Problem

**Raw spec validation** checks: *"Is this a valid OpenAPI document?"*
**Processed spec validation** checks: *"Is this internal representation valid?"*

These are **fundamentally different questions** with different answers.

## Real-World Impact

- ✅ **Raw spec**: Contains `$ref` + extra fields → Valid (extra fields ignored)
- ❌ **Processed spec**: Contains inlined schemas + same extra fields → Invalid (extra fields now in wrong context)

The same logical specification becomes invalid purely due to internal processing artifacts.

## The Fix

```python
def clone(self):
    return type(self)(copy.deepcopy(self._raw_spec))  # Always validate the original
```

This ensures `clone()` always produces a **validatable OpenAPI specification**, not a corrupted internal representation.

## Why This Matters

When Swagger UI calls `with_base_path()` → `clone()`, it expects to receive a **valid OpenAPI spec** that can be re-processed and displayed. Giving it a pre-processed spec violates this contract and causes validation failures.

---

*The fundamental principle: Only validate what the user provided, never validate internal processing artifacts.*

**Comment by chrisinmtown** [comment]
I think this might almost be a duplicate of PR #1889? That PR has been ignored by the maintainers for a long time. Now you have come along and done heroic work in figuring it out all over again! Until this fix or something like it gets merged, V3 remains fundamentally broken. I just don't know how to raise the priority of this in the queue.

**Comment by giuliohome** [comment]
Thank you @chrisinmtown, let me see if I can propose a slight modification to account for external refs

**Comment by giuliohome** [comment]
## New commit pushed with conditional cloning approach

I've updated the implementation to address the concerns about external references from PR #1889. The new approach:

- **For specs with only internal refs** (`#/components/...`): Uses `_raw_spec` to avoid validation artifacts from resolved schemas
- **For specs with external refs** (`file.yaml#/...`): Uses `_spec` to preserve resolved external content

This should maintain compatibility with the existing `relative_refs` tests while fixing the Swagger UI validation issue for specs with internal references only.

The `_has_only_internal_refs()` helper safely detects reference types by scanning for external file patterns (`.yaml`, `.json`) and non-fragment refs.

@RobbeSneyders This approach should address your previous concerns about test failures while still solving the core validation inconsistency issue.

**Comment by giuliohome** [comment]
**MyPy errors are unrelated to this change**

The MyPy failures appear to be pre-existing issues in the codebase:

```
connexion/middleware/abstract.py:57: error: "Specification" has no attribute "base_path"
connexion/middleware/abstract.py:60: error: "Specification" has no attribute "base_path"  
connexion/middleware/swagger_ui.py:73: error: "Specification" has no attribute "with_base_path"
connexion/middleware/main.py:422: error: "Type[Specification]" has no attribute "load"
```

These errors are in files (`middleware/abstract.py`, `middleware/swagger_ui.py`, `middleware/main.py`) that are completely unrelated to the changes in `spec.py`. The modifications in this PR only touch the `clone()` method and add a helper function `_has_only_internal_refs()` in the `Specification` class.

All formatting checks (flake8, isort, black) are now passing. The MyPy errors should be addressed separately as they represent existing type annotation issues in the middleware layer.

**Comment by giuliohome** [comment]
To be honest, I see all 802 tests passing (in my local WSL) even with my original one-line fix, but I added the conditional handler to be on the safe side.

## Testing Results

During testing, I verified that the `clone()` method is **always called with internal references only**, confirming that all **802 tests pass** using `_raw_spec`. 

The conditional handler was added for safety to handle potential external reference cases not covered by the current test suite.

### Key Findings:
- ✅ All 802 tests pass with the original one-line fix
- ✅ `_has_only_internal_refs()` returns `True` for all test cases
- ✅ No external references are used in the current test suite
- ✅ The conditional approach provides defensive coding for edge cases

**Comment by druizz90** [comment]
Good job @giuliohome.

Why isn't this PR merged?

**Comment by mathi123** [comment]
@Ruwann @RobbeSneyders Ping! Would you mind merging this?

**Comment by chrisinmtown** [comment]
> Why isn't this PR merged?

I am critically dependent on this package and have the same question.  FWIW I have personally never encountered a situation like this in the open-source world, with a popular (as far as I can tell) package and an active community, but disengaged maintainers. It's deeply concerning.

**Comment by RobbeSneyders** [comment]
Thanks @giuliohome 

I don't think the conditional handler is the way to go though. I assume the original issue addressed by #2002 was that the cloned spec was initialized without access to the `base_uri` of the original.

I believe the following should work:
- Store the `base_uri` in init: `self._base_uri = base_uri`
- Update clone to `return type(self)(copy.deepcopy(self._raw_spec), base_uri=self._base_uri)`

Could you try this and validate if your added test still passes?

**Comment by chrisinmtown** [comment]
hi @giuliohome I hope you don't mind me trying to follow the new suggestion and gather evidence about your PR. I made the change shown below, then ran tests. With this change in place, in my local the py3.9 .. py3.12 tests all pass including your new test case for external references. I think that's good news.

What about the py3.8 tests?  When I try `brew install python@3.8` I get `Error: python@3.8 has been disabled because it is deprecated upstream! It was disabled on 2024-10-14.` and that's keeping me from giving the suggested code a 100% pass result from here.

```
diff --git a/connexion/spec.py b/connexion/spec.py
index 2d95596..4206f02 100644
--- a/connexion/spec.py
+++ b/connexion/spec.py
@@ -77,6 +77,7 @@ class Specification(Mapping):
     operation_cls: t.Type[AbstractOperation]
 
     def __init__(self, raw_spec, *, base_uri=""):
+        self._base_uri = base_uri  # stash a reference
         self._raw_spec = copy.deepcopy(raw_spec)
         self._set_defaults(raw_spec)
         self._validate_spec(raw_spec)
@@ -204,12 +205,8 @@ class Specification(Mapping):
         return OpenAPISpecification(spec, base_uri=base_uri)
 
     def clone(self):
-        # Check if spec contains only internal refs (starting with #)
-        # For external refs, we need the processed spec to maintain resolved content
-        if self._has_only_internal_refs():
-            return type(self)(copy.deepcopy(self._raw_spec))
-        else:
-            return type(self)(copy.deepcopy(self._spec))
+        # use the base_uri stashed by __init__
+        return type(self)(copy.deepcopy(self._raw_spec), base_uri=self._base_uri)
 
     def _has_only_internal_refs(self):
         """Check if all $ref entries point to internal references only (starting with #)"""
```

**Comment by giuliohome** [comment]
Thanks for getting back to this. I’ve since moved on to another solution, so please feel free to continue on your end or close this PR.

**Comment by chrisinmtown** [comment]
@giuliohome please say, do you have a better solution to share with the community here?

**Comment by giuliohome** [comment]
Ah, I just meant we moved to another library for our use case. The proposed code should still work fine for this specific PR, but more broadly we needed something with a bit more active maintenance.

**Comment by chrisinmtown** [comment]
> Ah, I just meant we moved to another library for our use case.

Is there a direct competitor to connexion? Please share the name, I have similar concerns about the lack of active maintenance.

**Comment by RobbeSneyders** [comment]
Closing, replaced by https://github.com/spec-first/connexion/pull/2089
